content/blog/2023-12-11-A DSL for Metastability.md

```diff
@@ -0,0 +1,237 @@
+
+There is also an entity called the environment, which is responsible for *triggers*.
+The environment is not expressed by the DSL, and is assumed to operate based on pure choice: each "action" for the environment is just an arbitrary state transformation subject to the limitations imposed on the environment.
+These limitations are expressed by specifying the variables that the environment can touch.
+For instance, the code for some agent can be wrapped with a giant branching action with a guard $\texttt{not crashed}$, and a $\texttt{nop}$ for the else clause.
+The environment can then choose to crash and uncrash the agent however it likes.
+
+Based on these, a general execution of the system is a repeating cycle of the following:
+1. The environment takes an action.
+2. The system as a whole takes an action, i.e., each agent takes a big step.
+3. The network handles the messages.
+
+## Example
+Time for an example.
+This example is from an old paper that was trying to study a real-world phenomenon in routing networks.
+We have identified the phenomenon as being metastable, which is good news for the DSL's ability to express systems in which metastability is observed.
+There are other examples as well (the retry storm and a membership scenario from an industry setup), but for the sake of brevity I have not included them here.
+Note that I have heavily used sugaring, as you can imaging it is not pleasant to write actual code in the specified grammar.
+The sugaring should be self-explanatory.
+Let us first see the code, and then I will explain the scenario.
+
+```
+### Receive
+while not inq.is_full:
+    m = take();
+    put(inq, m);
+
+### Process
+if not is_processing {
+    t_t = t_t + 1;
+} else {
+    t_p = t_p + 1;
+}
+if t_p == T_p {
+    is_processing = false;
+    t_p = 0;
+    T_p = T_p0;
+} else {
+    nop;
+}
+if t_t == T_t {
+    is_processing = true;
+    t_t = 0
+    r = {type: "routing message", payload: "-", destination: BROADCAST};
+    if not outq.is_full {
+        put(outq, r);
+    } else {
+        nop;
+    }
+}
+while not inq.is_empty {
+    m = get(inq);
+    if m.type == "routing message" {
+        if is_processing {
+            T_p = T_p + 1;
+        } else {
+            nop;
+        }
+    } else if m.type == "update" {
+        is_processing = true;
+        t_t = 0
+        r = {type: "routing message", payload: "-", destination: BROADCAST};
+        if not outq.is_full {
+            put(outq, r);
+        } else {
+            nop;
+        }
+    } else {
+        nop;
+    }
+}
+
+### Send
+while not outq.is_empty:
+    m = get(outq);
+    give(m);
+```
+First, some notes:
+1. $\texttt{take}$ and $\texttt{give}$ are keywords to interact with the network. I did not write them in the grammar as I am still not sure how to treat them. Informally, $\texttt{take}$ receives the next request to be delivered by the network, and $\texttt{give}$ gives a request from the output queue to the network.
```

Comment:
This probably doesn't matter at all, but FWIW it is possible to typeset code in Markdown with backticks. So `` `take` `` would be a suitable alternative to `$\texttt{take}$` that avoids the need for the LaTeX renderer.